rfd/0048-war-dialler-node-tracker.md

```diff
@@ -0,0 +1,199 @@
+The second RFD proposes a new proxy-to-proxy tunneling system, and can be tracked [here](insert rfd link here)
+
+## Why
+
+The current behavior of node agents dialing over a reverse tunnel are required to connect to every proxy instance in the cluster. This allows a client to connect to a node through any proxy but it also causes other problems when running proxies behind a load balancer like in our cloud environment. These problems include:
+
+- Ephemeral port exhaustion between a NAT gateway and load balancer. This limits the number of nodes that can be connected behind a single NAT gateway. As more proxies are added fewer nodes are able to connect.
+- Thundering herd when adding, removing, or restarting a proxy. Node agents retry connecting until they randomly get balanced to the desired proxy. The more node agents connected the worse this problem becomes.
+
+Both these issues are mitigated by changing the node agent behavior to connect to a single proxy. Ephemeral port exhaustion is no longer tied to the number of proxies and node agents no longer need to retry until they connect to a specific proxy.
+
+## Details
+
+Here's a small example comparing how proxies would interact with the node tracker:
+
+```
+	Original behaviour:
+                       ┌──────┐
+   ┌───connection─────>│proxy1│<──reverse-tunnel──┐
+   │                   └──────┘                   │
+┌──┴───┐                                        ┌─┴──┐
+│client│                                        │node│
+└──┬───┘                                        └─┬──┘
+   │                   ┌──────┐                   │
+   └───connection─────>│proxy2│<──reverse-tunnel──┘
+                       └──────┘
+	A client can reach the requested node through any proxy because the node has a reverse tunnel open to all proxies in a cluster.
+
+
+
+    New behaviour:
+                         ┌──────┐         ┌────────────┐
+   ┌─(1)─connection-1───>│proxy1├──(2)───>│node tracker│
+   │                     └──╥───┘         └────────────┘
+┌──┴───┐                    ║
+│client│               (3) tunnel
+└──┬───┘                    ║
+   │                     ┌──╨───┐                    ┌────┐
+   └────connection-2────>│proxy2│<───reverse-tunnel──┤node│
+                         └──────┘                    └────┘
+
+	Case 1 - connection 1:
+	1- A client connects to proxy-1 that can't reach the requested node.
+	2- Proxy-1 sends a lookup request to the node tracker asking which proxy is connected to the requested node. The node tracker replies with the address of proxy-2.
+	3- Proxy-1 opens a tunnel (RFD 49?) to proxy-2 and asks proxy-2 to dial the requested node.
+
+	Case 2 - connection 2:
+	nothing changes when a client connects to a proxy that has got a direct connection to the requested node.
+```
+
+### Scope
+
+The first iteration of the node tracker service will be a barebones implementation of an in memory cache represented by a simple golang map.
```

Comment:
What are cache expiration policies?

Reply:
It can be configured in the config of the node tracker but it defaults to
https://github.com/gravitational/teleport/blob/ab35f11ce540d15faf68b42380428c83ffa7d2e1/api/defaults/defaults.go#L64